drivers/qcommsimmodem/radio-settings.c

```diff
@@ -0,0 +1,272 @@
+/*
+ *
+ *  oFono - Open Source Telephony
+ *
+ *  Copyright (C) 2014  Canonical Ltd.
+ *  Copyright (C) 2015 Ratchanan Srirattanamet
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 2 as
+ *  published by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include <glib.h>
+
+#include <ofono/log.h>
+#include <ofono/modem.h>
+#include <ofono/radio-settings.h>
+
+#include "gril.h"
+#include "grilrequest.h"
+#include "grilreply.h"
+
+#include "drivers/rilmodem/radio-settings.h"
+#include "drivers/rilmodem/rilutil.h"
+#include "qcom_msim_modem.h"
+#include "qcom_msim_constants.h"
+
+struct qcom_msim_pending_pref_setting {
+	struct ofono_radio_settings *rs;
+	int pref;
+	int pending_gsm_pref_remaining;
+	struct cb_data *cbd;
+};
+
+struct qcom_msim_set_2g_rat {
+	struct ofono_radio_settings *rs;
+	struct qcom_msim_pending_pref_setting *pps;
+};
+
+static struct ofono_radio_settings *multisim_rs[QCOMMSIM_NUM_SLOTS_MAX];
+static int multisim_num_slot;
```

Comment:
This really should be plural... *multisim_num_slots*, as that's what it is, the number of slots.